### PR TITLE
[libc++] Deprecated `shared_ptr` Atomic Access APIs as per P0718R2 & Implemented P2869R3: Remove Deprecated `shared_ptr` Atomic Access APIs from C++26

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Implements https://wg21.link/P2869R4
Implements deprecations as per https://wg21.link/P0718R2

### CODE REVIEW

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for working on this! 
While reviewing your patch I noticed we have not implemented [P0718R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0718r2.html). This paper deprecates `shared_ptr` Atomic Access APIs. I don't like to remove this feature when we don't issue deprecation messages. Can you implement [P0718R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0718r2.html) too? I'm fine to have it in the same patch. If you do it in this patch please update the title and the description.

**Comment by H-G-Hristov** [comment]
> Thanks for working on this! While reviewing your patch I noticed we have not implemented [P0718R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0718r2.html). This paper deprecates `shared_ptr` Atomic Access APIs. I don't like to remove this feature when we don't issue deprecation messages. Can you implement [P0718R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0718r2.html) too? I'm fine to have it in the same patch. If you do it in this patch please update the title and the description.

@mordante Thank you for the review. I implemented the deprecations as per [P0718R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0718r2.html) The remaining parts of the paper are not a subject of this patch.

File: libcxx/docs/Status/Cxx20Papers.csv [hunk 1]
```diff
@@ -12,7 +12,7 @@
 "`P0600R1 <https://wg21.link/P0600R1>`__","LWG","nodiscard in the Library","Albuquerque","|Complete|","16.0"
 "`P0616R0 <https://wg21.link/P0616R0>`__","LWG","de-pessimize legacy <numeric> algorithms with std::move","Albuquerque","|Complete|","12.0"
 "`P0653R2 <https://wg21.link/P0653R2>`__","LWG","Utility to convert a pointer to a raw pointer","Albuquerque","|Complete|","6.0"
-"`P0718R2 <https://wg21.link/P0718R2>`__","LWG","Atomic shared_ptr","Albuquerque","",""
+"`P0718R2 <https://wg21.link/P0718R2>`__","LWG","Atomic shared_ptr","Albuquerque","|Partial|","19.0"
```
Comment (mordante) [inline libcxx/docs/Status/Cxx20Papers.csv:15]: This adds a link to the note you created.
In general we only add a version when it's complete, feel free to add the version to the note.
```suggestion
"`P0718R2 <https://wg21.link/P0718R2>`__","LWG","Atomic shared_ptr","Albuquerque","|Partial| [#note-P0718]_",""
```

File: libcxx/test/std/utilities/memory/util.smartptr/util.smartptr.shared.atomic/atomic_compare_exchange_strong.verify.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+//===----------------------------------------------------------------------===//
```
Comment (mordante) [inline libcxx/test/std/utilities/memory/util.smartptr/util.smartptr.shared.atomic/atomic_compare_exchange_strong.verify.cpp:1]: These deprecated tests are typically in `libcxx/test/libcxx/depr/depr.XXX` where XXX is the section name in the Standard. They typically have a cxxYY too in their name. (We're not very consistent in this naming but let's match that generic direction.)

**Review by mordante — APPROVED** [review]
LGTM modulo some minor nits.

**Comment by H-G-Hristov** [comment]
> LGTM modulo some minor nits.

Thank you!

**Comment by nico** [comment]
As mentioned before, it'd be very useful to have opt-out flags for deprecations.

https://github.com/llvm/llvm-project/pull/80542#issuecomment-2045532450 claims that "deprecation warnings are introduced in newer standards only", but in our experience that just isn't true (and isn't true here). The MSVC STL does have an opt-out macro per deprecated thing. (@ldionne)

We're once again in the situation that we can't update libc++ before cleaning up this (and possibly other) deprecation warnings. It would be considerably easier if we could update (with a define to disable this deprecation warning) and the fix the warning asynchronously.

(I didn't notice this earlier because it took us 5 weeks to fix the last thing that we couldn't fix asynchronously.)

**Comment by Zingam** [comment]
AFAIK that comment states that `std::err` is applied as DR, which means it is applied to previous standards retroactively. 
We don't do that to atomic access APIs here, which are deprecated in C++20 and that C++20 paper (P0718R2) has not been implemented up until now.
If we have a policy to add opt-out flags I'd be happy to implement them.

**Comment by nico** [comment]
We're trying to use the replacement over in https://github.com/google/perfetto/issues/785 .  https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2869r2.pdf suggests to use `std::atomic<std::shared_ptr<...>>`. But we're getting errors that `_Atomic cannot be applied to type 'std::shared_ptry<X>' which is not trivially copiable`.

Maybe we're holding something wrong, but https://libcxx.llvm.org/FeatureTestMacroTable.html says "__cpp_lib_atomic_shared_ptr unimplemented". Which suggests that the recommended replacement doesn't work in ilbc++ yet.

Is that correct?

If so, I'll suggest that adding this deprecation is premature.

**Comment by zmodem** [comment]
@mordante and @H-G-Hristov: ping? Have you had a chance to consider nico's [comment](https://github.com/llvm/llvm-project/pull/87111#issuecomment-2112740039) above?

**Pull request merged**